### ISSUE TITLE
stripe.api_requestor._api_encode ambiguously encodes list of dicts

### ISSUE THREAD

**Issue by nickbaum** [stripe/stripe-python#273]
I'm filing this bug to follow up on a support conversation with Thomas and Armand, who were able to reproduce this issue.

I dug into how the request arguments are converted, and it seems stripe.api_requestor._api_encode encodes the items in an ambiguous way:

```
>>> items = [{'quantity': 1, 'parent': 'black', 'type': 'sku'}, {'description': 'Gift credit', 'type': 'discount', 'amount': -3900, 'currency': 'usd'}] 
>>> list(stripe.api_requestor._api_encode({"items": items}))
[('items[][quantity]', 1), ('items[][type]', 'sku'), ('items[][parent]', 'black'), ('items[][type]', 'discount'), ('items[][description]', 'Gift credit'), ('items[][currency]', 'usd'), ('items[][amount]', -3900)]
```

This can result in the Python library submitting incorrect data to the API.

From Armand:

> Armand here, jumping in for Thomas, I hope you don't mind! I've spent some time investigating the issue and I believe you're correct: lists of hashes are encoded in an ambiguous way. This, coupled with the fact that Python hashes (dicts) are unordered, can result in inconsistent behavior.
> 
> One possible workaround would be to use OrderedMap and make sure to insert the `type` attribute first in each item's dict. On the API's end, when a key appears twice in the same list, it will be interpreted as a new element in the list.
> 
> Using OrderedMaps is a bit unwieldy though, so a simpler workaround is to replace the list by a hash with incrementing numerical keys, e.g.:
> 
> ```
> >>> items = {0: {'quantity': 1, 'parent': 'black', 'type': 'sku'}, 1: {'description': 'Gift credit', 'type': 'discount', 'amount': -3900, 'currency': 'usd'}}
> >>> list(stripe.api_requestor._api_encode({"items": items}))
> [('items[1][amount]', -3900), ('items[1][description]', 'Gift credit'), ('items[1][type]', 'discount'), ('items[1][currency]', 'usd'), ('items[0][quantity]', 1), ('items[0][parent]', 'black'), ('items[0][type]', 'sku')]
> ```
> 
> This removes the ambiguity and will be decoded correctly on the API's end.
> 
> In any case, this is probably something that should be handled in the bindings themselves. I will forward the details to the bindings maintainers and discuss the best way to fix this issue.

**Comment by brandur** [comment]
Thank you for logging this! We're looking into the best way to get this fixed.

**Comment by brandur-stripe** [comment]
Sorry about the very slow follow up on this one, but Olivier just committed a fix as of #342 whereby `items` is encoded as a list with explicit indexes. The patch is available as part of 1.63.1. I'm closing this issue out for now.

**Comment by nickbaum** [comment]
🙌🏻
